### PR TITLE
resource instance support gpdb

### PR DESCRIPTION
**Pull request by ouyangyewei**
1. resoruce instance support gpdb
2. add documents for resource.
3. add gpdb golang sdk.

```
TF_ACC=1 go test ./alicloud -v -run=TestAccAlicloudGpdbInstance_classic -timeout=300m
=== RUN   TestAccAlicloudGpdbInstance_classic
--- PASS: TestAccAlicloudGpdbInstance_classic (631.04s)
PASS
ok      github.com/terraform-providers/terraform-provider-alicloud/alicloud     631.101s



 TF_ACC=1 go test ./alicloud -v -run=TestAccAlicloudGpdbInstance_vpc -timeout=300m
=== RUN   TestAccAlicloudGpdbInstance_vpc
--- PASS: TestAccAlicloudGpdbInstance_vpc (717.70s)
PASS
ok      github.com/terraform-providers/terraform-provider-alicloud/alicloud     717.763s
```

### CODE REVIEW

File: alicloud/resource_alicloud_gpdb_instance.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+		if err != nil {
+			return nil, WrapError(err)
+		}
+
+		if request.ZoneId == "" {
+			request.ZoneId = vsw.ZoneId
+		} else if strings.Contains(request.ZoneId, MULTI_IZ_SYMBOL) {
+			zoneStr := strings.Split(strings.SplitAfter(request.ZoneId, "(")[1], ")")[0]
+			if !strings.Contains(zoneStr, string([]byte(vsw.ZoneId)[len(vsw.ZoneId)-1])) {
+				return nil, WrapError(fmt.Errorf("The specified vswitch %s isn't in the multi zone %s.", vsw.VSwitchId, request.ZoneId))
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_gpdb_instance.go]: fmt.Errorf()请替换成Error()方法

File: alicloud/resource_alicloud_gpdb_instance.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+			}
+		} else if request.ZoneId != vsw.ZoneId {
+			return nil, WrapError(fmt.Errorf("The specified vswitch %s isn't in the zone %s.", vsw.VSwitchId, request.ZoneId))
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_gpdb_instance.go]: 同上

File: alicloud/resource_alicloud_gpdb_instance_test.go [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheckWithRegions(t, false, connectivity.GpdbClassicNoSupportedRegions)
+		},
+		IDRefreshName: resourceId,
+		Providers:     testAccProviders,
+		CheckDestroy:  testAccCheckGpdbInstanceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testGpdbInstance_classic_basic,
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_gpdb_instance_test.go]: testcase 请使用新的格式，可以参考dns_record的testcase，并且添加import的testcase

File: alicloud/service_alicloud_gpdb.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		return client.DescribeDBInstanceAttribute(request)
+	})
+
+	response, _ := raw.(*gpdb.DescribeDBInstanceAttributeResponse)
+	if err != nil {
+		return instanceAttribute, WrapErrorf(err, DefaultErrorMsg, id, request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+
+	addDebug(request.GetActionName(), response)
+	if response == nil || len(response.Items.DBInstanceAttribute) == 0 {
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_gpdb.go]: 这里不用判断response == nil，可以删除

File: alicloud/service_alicloud_gpdb.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+		return instanceAttribute, WrapErrorf(Error(GetNotFoundMessage("Gpdb Instance", id)), NotFoundMsg, AlibabaCloudSdkGoERROR)
+	}
+
+	return response.Items.DBInstanceAttribute[0], nil
+}
+
+func (s *GpdbService) GetSecurityIps(instanceId string) ([]string, error) {
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_gpdb.go]: instanceId改为id，统一使用id作为参数名

File: alicloud/service_alicloud_gpdb.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+func (s *GpdbService) DescribeGpdbInstance(id string) (instanceAttribute gpdb.DBInstanceAttribute, err error) {
+	request := gpdb.CreateDescribeDBInstanceAttributeRequest()
+	request.DBInstanceId = id
+	raw, err := s.client.WithGpdbClient(func(client *gpdb.Client) (interface{}, error) {
+		return client.DescribeDBInstanceAttribute(request)
+	})
+
+	response, _ := raw.(*gpdb.DescribeDBInstanceAttributeResponse)
+	if err != nil {
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_gpdb.go]: 这里如果err为notfound，是否应该判断一下

File: website/docs/r/gpdb_instance.html.markdown [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+# alicloud\_gpdb\_instance
+
+Provides a AnalyticDB for PostgreSQL instance resource supports replica set instances only. the AnalyticDB for PostgreSQL provides stable, reliable, and automatic scalable database services. 
+You can see detail product introduction [here](https://www.alibabacloud.com/help/doc-detail/35387.htm)
+
+-> **NOTE:**  The following regions don't support create Classic network Gpdb instance.
+[`ap-southeast-2`,`ap-southeast-3`,`ap-southeast-5`,`ap-south-1`,`me-east-1`,`ap-northeast-1`,`eu-west-1`,`us-east-1`,`eu-central-1`,`cn-shanghai-finance-1`,`cn-shenzhen-finance-1`,`cn-hangzhou-finance`]
+
+-> **NOTE:**  Create instance or change instance would cost 10~15 minutes. Please make full preparation.
+
```
Comment (YY-tomorrow) [inline website/docs/r/gpdb_instance.html.markdown]: 加一个note，从哪个版本后开始支持

File: alicloud/resource_alicloud_gpdb_instance.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+		})
+		if err != nil {
+			if gpdbService.NotFoundGpdbInstance(err) {
+				return resource.NonRetryableError(err)
+			}
+			return resource.RetryableError(err)
+		}
+		addDebug(request.GetActionName(), raw)
+		return nil
+	})
```
[the remark this answers is not in the document]
Comment (ouyangyewei) [reply alicloud/resource_alicloud_gpdb_instance.go]: 这个retry的逻辑暂时继续保留

File: alicloud/resource_alicloud_gpdb_instance.go [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+	request.DBInstanceGroupCount = Trim(d.Get("instance_group_count").(string))
+	request.Engine = Trim(d.Get("engine").(string))
+	request.EngineVersion = Trim(d.Get("engine_version").(string))
+
+	// Instance NetWorkType
+	request.InstanceNetworkType = string(Classic)
+	if request.VSwitchId != "" {
+		// check vswitchId in zone
+		vpcService := VpcService{client}
+		vsw, err := vpcService.DescribeVSwitch(request.VSwitchId)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_gpdb_instance.go]: vsw改成object，service方法返回的值统一使用object

File: alicloud/service_alicloud_gpdb.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+		if IsExceptedErrors(err, []string{InvalidGpdbInstanceIdNotFound, InvalidGpdbNameNotFound}) {
+			err = WrapErrorf(err, NotFoundMsg, AliyunLogGoSdkERROR)
+		} else {
+			err = WrapErrorf(err, DefaultErrorMsg, id, "GetFunction", AliyunLogGoSdkERROR)
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_gpdb.go]: 25行和27行的AliyunLogGoSdkERROR ，应该改成AlibabaCloudSdkGoERROR

File: alicloud/resource_alicloud_gpdb_instance.go [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+			}
+		}
+		addDebug(request.GetActionName(), raw)
+		return nil
+	})
+	if err != nil {
+		if gpdbService.NotFoundGpdbInstance(err) {
+			return nil
+		}
+		return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), GpdbGoSdk)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_gpdb_instance.go]: AlibabaCloudSdkGoERROR

File: alicloud/errors.go [hunk 12]
```diff
@@ -548,6 +553,7 @@ const (
 	AliyunDatahubSdkGo     = ErrorSource("[SDK aliyun-datahub-sdk-go ERROR]")
 	AliyunOssGoSdk         = ErrorSource("[SDK aliyun-oss-go-sdk ERROR]")
 	FcGoSdk                = ErrorSource("[SDK fc-go-sdk ERROR]")
+	GpdbGoSdk              = ErrorSource("[SDK gpdb-go-sdk ERROR]")
```
Comment (YY-tomorrow) [inline alicloud/errors.go]: this is no need

File: alicloud/service_alicloud_gpdb.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+func (s *GpdbService) DescribeGpdbSecurityIps(instanceId string) (ips []gpdb.DBInstanceIPArray, err error) {
+	request := gpdb.CreateDescribeDBInstanceIPArrayListRequest()
+	request.DBInstanceId = instanceId
+
+	raw, err := s.client.WithGpdbClient(func(client *gpdb.Client) (interface{}, error) {
+		return client.DescribeDBInstanceIPArrayList(request)
+	})
+	if err != nil {
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_gpdb.go]: 这里notfound的判断

File: alicloud/service_alicloud_gpdb.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+			err = WrapErrorf(err, NotFoundMsg, AlibabaCloudSdkGoERROR)
+		} else {
+			err = WrapErrorf(err, DefaultErrorMsg, id, "GetFunction", AlibabaCloudSdkGoERROR)
+		}
+		return
+	}
+
+	addDebug(request.GetActionName(), response)
+	if len(response.Items.DBInstanceAttribute) == 0 {
+		return instanceAttribute, WrapErrorf(Error(GetNotFoundMessage("Gpdb Instance", id)), NotFoundMsg, AlibabaCloudSdkGoERROR)
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_gpdb.go]: AlibabaCloudSdkGoERROR 应该为ProviderERROR

File: alicloud/service_alicloud_gpdb.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+	var finalIps []string
+	if len(ipsMap) > 0 {
+		for key := range ipsMap {
+			finalIps = append(finalIps, key)
+		}
+	}
+	return finalIps, nil
+}
+
+func (s *GpdbService) DescribeGpdbSecurityIps(instanceId string) (ips []gpdb.DBInstanceIPArray, err error) {
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_gpdb.go]: instanceId 换成 id

File: alicloud/data_source_alicloud_zones.go [hunk 16]
```diff
@@ -233,6 +237,29 @@ func dataSourceAlicloudZonesRead(d *schema.ResourceData, meta interface{}) error
 			}
 		}
 	}
+	if strings.ToLower(Trim(resType)) == strings.ToLower(string(ResourceTypeGpdb)) {
+		request := gpdb.CreateDescribeRegionsRequest()
+		raw, err := client.WithGpdbClient(func(gpdbClient *gpdb.Client) (interface{}, error) {
+			return gpdbClient.DescribeRegions(request)
+		})
+		if err != nil {
+			return WrapErrorf(err, DataDefaultErrorMsg, "aliCloud_zones", request.GetActionName(), AlibabaCloudSdkGoERROR)
```
Comment (xiaozhu36) [inline alicloud/data_source_alicloud_zones.go]: `aliCloud_zones ` -> `alicloud_zones`

File: alicloud/resource_alicloud_gpdb_instance.go [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Optional: true,
+			},
+			"vswitch_id": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Optional: true,
+				Computed: true,
+			},
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_instance.go]: I think the field `instance_network_type ` and `vpc_id` are redundant and them can be get by `vswitch_id`

File: alicloud/resource_alicloud_gpdb_instance_test.go [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+)
+
+// Acceptance Test
+// Create by yewei.oyyw@alibaba-inc.com on 2019-05-31
+
+func init() {
+	resource.AddTestSweepers("alicloud_gpdb_instance", &resource.Sweeper{
+		Name: "alicloud_gpdb_instance",
+		F:    testSweepGpdbInstances,
+	})
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_instance_test.go]: The sweeper test should be a dependence of `alicloud_vswitch's` sweeper.

File: alicloud/service_alicloud_gpdb.go [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+	raw, err := s.client.WithGpdbClient(func(client *gpdb.Client) (interface{}, error) {
+		return client.DescribeDBInstanceAttribute(request)
+	})
+
+	response, _ := raw.(*gpdb.DescribeDBInstanceAttributeResponse)
+	if err != nil {
+		if IsExceptedErrors(err, []string{InvalidGpdbInstanceIdNotFound, InvalidGpdbNameNotFound}) {
+			err = WrapErrorf(err, NotFoundMsg, AlibabaCloudSdkGoERROR)
+		} else {
+			err = WrapErrorf(err, DefaultErrorMsg, id, "GetFunction", AlibabaCloudSdkGoERROR)
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_gpdb.go]: `GetFunction` -> `request.GetActionName()`

File: alicloud/service_alicloud_gpdb.go [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+	return false
+}
+
+// Wait for instance to given status
+func (s *GpdbService) WaitForGpdbInstance(id string, status Status, timeout int) error {
+	deadline := time.Now().Add(time.Duration(timeout) * time.Second)
+	for {
+		instance, err := s.DescribeGpdbInstance(id)
+		if err != nil {
+			if s.NotFoundGpdbInstance(err) {
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_gpdb.go]: There is no need `s.NotFoundGpdbInstance` and it just `NotFoundError(err)` is ok.

File: alicloud/service_alicloud_gpdb.go [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	response := raw.(*gpdb.ModifySecurityIpsResponse)
+	addDebug(request.GetActionName(), response)
+	if err := s.WaitForGpdbInstance(id, Running, DefaultTimeoutMedium); err != nil {
+		return WrapError(err)
+	}
+	return nil
+}
+
+func (s *GpdbService) NotFoundGpdbInstance(err error) bool {
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_gpdb.go]: The method `NotFoundGpdbInstance` is redundant. Please remove it.

File: alicloud/resource_alicloud_gpdb_instance.go [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+			if IsExceptedErrors(err, []string{InvalidGpdbInstanceStatus}) {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		addDebug(request.GetActionName(), raw)
+		return nil
+	})
+	if err != nil {
+		if gpdbService.NotFoundGpdbInstance(err) {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_gpdb_instance.go]: There is no need to use `gpdbService.NotFoundGpdbInstance(err)` just `IsExceptedErrors(err, []string{InvalidGpdbInstanceIdNotFound, InvalidGpdbNameNotFound})` is ok.

File: alicloud/connectivity/regions.go [hunk 23]
```diff
@@ -76,3 +77,4 @@ var SlbClassicNoSupportedRegions = []Region{APNorthEast1, APSouthEast2, APSouthE
 var NasNoSupportedRegions = []Region{Qingdao, APSouth1, APSouthEast3, APSouthEast5}
 var OssVersioningSupportedRegions = []Region{APSouth1}
 var OssSseSupportedRegions = []Region{Qingdao, Hangzhou, Beijing, Shanghai, Shenzhen, Hongkong, APNorthEast1, APSouth1, USEast1}
+var GpdbClassicNoSupportedRegions = []Region{APSouthEast2, APSouthEast3, APSouthEast5, APSouth1, USEast1, USWest1, APNorthEast1, EUCentral1}
```
Comment (xiaozhu36) [inline alicloud/connectivity/regions.go]: The supported and unsupported regions should be applied to the acceptance test cases.

File: alicloud/service_alicloud_gpdb.go (hunk 20 above)
Comment (ouyangyewei) [reply alicloud/service_alicloud_gpdb.go]: NotFoundError(err) do not contains gpdb error codes.

**Pull request merged**